src/store/store.js

```diff
@@ -0,0 +1,265 @@
+
+		listeners.forEach((_, listener) => {
+			listener(null, error);
+		});
+	}
+
+	setError(entityId, token, error) {
+		const lowerCaseEntityId = entityId.toLowerCase();
+		const tokenCache = token.toString();
+		this._entities.get(tokenCache).set(lowerCaseEntityId, {
+			status: 'error',
+			entity: null,
+			error: error,
+			request: null
+		});
+		this.notifyError(entityId, token, error);
+	}
+
+	_setFetchingState(entityId, token) {
+		const lowerCaseEntityId = entityId.toLowerCase();
+		const tokenCache = token.toString();
+		this._entities.get(tokenCache).set(lowerCaseEntityId, {
+			status: 'fetching',
+			entity: null
+		});
+	}
+
+	_initContainer(map, entityId, token, init) {
+		const lowerCaseEntityId = entityId.toLowerCase();
+		const tokenCache = token.toString();
+		if (!map.has(tokenCache)) {
+			map.set(tokenCache, new Map());
+		}
+		const entityMap = map.get(tokenCache);
+		if (init && !entityMap.has(lowerCaseEntityId)) {
+			entityMap.set(lowerCaseEntityId, init);
+		}
+		return entityMap.get(lowerCaseEntityId);
+	}
+
+	async _handleCachePriming(token, response) {
+		const linkHeaderValues = response.headers && response.headers.get('Link');
+		if (!linkHeaderValues) {
+			return;
+		}
+
+		const cachePrimers = this._parseLinkHeader(linkHeaderValues)
+			.filter(function(link) {
+				return link.rel.indexOf('https://api.brightspace.com/rels/cache-primer') !== -1;
+			});
+
+		if (cachePrimers.length === 0) {
+			return;
+		}
+
+		await Promise.all(cachePrimers.map((cachePrimer) => {
+			return this.fetch(cachePrimer.href, token, true);
+		}));
+	}
+
+	// parse a Link header
+	//
+	// Link:<https://example.org/.meta>; rel=meta
+	//
+	// var r = parseLinkHeader(xhr.getResponseHeader('Link');
+	// r['meta'] outputs https://example.org/.meta
+	//
+	_parseLinkHeader(links) {
+		const linkexp = /<[^>]*>\s*(\s*;\s*[^\(\)<>@,;:"\/\[\]\?={} \t]+=(([^\(\)<>@,;:"\/\[\]\?={} \t]+)|("[^"]*")))*(,|$)/g; // eslint-disable-line no-useless-escape
+		const paramexp = /[^\(\)<>@,;:"\/\[\]\?={} \t]+=(([^\(\)<>@,;:"\/\[\]\?={} \t]+)|("[^"]*"))/g; // eslint-disable-line no-useless-escape
+
+		const matches = links.match(linkexp);
+		const _links = [];
+		for (let i = 0; i < matches.length; i++) {
+			const split = matches[i].split('>');
+			const href = split[0].substring(1);
+			_links.push({
+				href
+			});
+			const ps = split[1];
+			const s = ps.match(paramexp);
+			for (let j = 0; j < s.length; j++) {
+				const p = s[j];
+				const paramsplit = p.split('=');
+				const name = paramsplit[0];
+				const val = paramsplit[1].replace(/["']/g, '');
+				if (name === 'rel') {
+					const relsplit = val.split(' ');
+					_links[i][name] = relsplit;
+				} else {
+					_links[i][name] = val;
+				}
+			}
+		}
+		return _links;
+	}
+}
+
+window.D2L = window.D2L || {};
+window.D2L.SirenSdk = window.D2L.SirenSdk || {};
+window.D2L.SirenSdk.EntityStore = window.D2L.SirenSdk.EntityStore || new EntityStore(window.d2lfetch);
```

Comment:
Yeah, I would like to get away from using the window globals. I wasn't sure how to make sure it was global to every component on the page. 

I'd also like to split this off at some point so the user can define where hold the data. That way we can use the `siren-sdk` on backend services too. 

Exporting the methods below this does remove the user from accessing or requiring knowledge that the store is stored on the window.

Reply:
Is the issue here that you want the entity store to be a singleton? Would there ever be a use case where you'd want multiple entity stores?

Reply:
I figured I'd make the class not required to be a singleton. For the use case on the UI we will want it to be pretty much a singleton.

I can wrap this in a factory. Didn't really know when to call that in the UI though. I assuming it would be called in EntityFactory but then we will just be calling it on that input.

Reply:
You could use lazy initialization to create the singleton (e.g. create it the first time someone tries to access it). Alternatively, you could instantiate your singleton in the global scope of this file. This means the singleton gets created when the file is first loaded. Assuming `import` works similar to `require`, then there is caching so a file only gets loaded once even if it's imported multiple places. This would guard against multiple singletons being created.

Reply:
For example:
```
class EntityStore { ... }
export const singleton = new EntityStore( ... );
```

Reply:
I'd tread carefully assuming that the module loaders are uniform.  Whatwg defines the module map as such https://html.spec.whatwg.org/multipage/webappapis.html#module-map, but you'd have to also ensure that the transpiling to es5 also behaves the same.

It's possible that we don't care about it though, but it would feel slightly safer to factory method with a cache on windoworglobal

Reply:
So making a factory sounds like the best way. So I'll do that for sure.